src/sentry/metrics/minimetrics.py

```diff
@@ -0,0 +1,323 @@
+import threading
+import time
+import zlib
+from threading import Lock, Thread
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generic,
+    List,
+    Literal,
+    Optional,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from sentry.metrics.base import MetricsBackend, Tags
+from sentry.utils import metrics
+
+# The thread local instance must be initialized globally in order to correctly use the state.
+thread_local = threading.local()
+
+
+T = TypeVar("T")
+
+
+MetricUnit = Literal[
+    None,
+    "nanosecond",
+    "microsecond",
+    "millisecond",
+    "second",
+    "minute",
+    "hour",
+    "day",
+    "week",
+    "bit",
+    "byte",
+    "kilobyte",
+    "kibibyte",
+    "mebibyte",
+    "gigabyte",
+    "terabyte",
+    "tebibyte",
+    "petabyte",
+    "pebibyte",
+    "exabyte",
+    "exbibyte",
+    "ratio",
+    "percent",
+]
+
+
+class Metric(Generic[T]):
+    def add(self, value: T) -> None:
+        raise NotImplementedError()
+
+    def serialize_value(self) -> Any:
+        raise NotImplementedError()
+
+
+class CounterMetric(Metric[float]):
+    __slots__ = ("value",)
+
+    def __init__(self) -> None:
+        self.value = 0.0
+
+    def add(self, value: float) -> None:
+        self.value += value
+
+    def serialize_value(self) -> Any:
+        return self.value
+
+
+class GaugeMetric(Metric[float]):
+    __slots__ = ("min", "max", "sum", "count", "last")
+
+    def __init__(self) -> None:
+        self.min = float("inf")
+        self.max = float("-inf")
+        self.sum = 0.0
+        self.count = 0.0
+        self.last = float("nan")
+
+    def add(self, value: float) -> None:
+        self.min = min(self.min, value)
+        self.max = max(self.max, value)
+        self.last = value
+        self.count += 1
+        self.sum += value
+
+    def serialize_value(self) -> Any:
+        return {
+            "min": self.min,
+            "max": self.max,
+            "last": self.last,
+            "sum": self.sum,
+            "count": self.count,
+        }
+
+
+class DistributionMetric(Metric[float]):
+    __slots__ = ("value",)
+
+    def __init__(self) -> None:
+        self.value: List[float] = []
+
+    def add(self, value: float) -> None:
+        self.value.append(value)
+
+    def serialize_value(self) -> Any:
+        return self.value
+
+
+class SetMetric(Metric[Union[str, int]]):
+    __slots__ = ("value",)
+
+    def __init__(self) -> None:
+        self.value: Set[Union[str, int]] = set()
+
+    def add(self, value: Union[str, int]) -> None:
+        self.value.add(value)
+
+    def serialize_value(self) -> Any:
+        def _hash(x: Any) -> int:
+            if isinstance(x, str):
+                return zlib.crc32(x.encode("utf-8")) & 0xFFFFFFFF
```

Comment:
You can add a note here that this is currently required because of the bucket protocol in Relay. We would allow ingesting strings if we stick to the JSON protocol, or if we use the text protocol we could send tag values directly.